File: src/dataset/python/bowling_exercism-ThomasZumsteg/claude/ClaudeSonnet4_bowling_v2.py
class BowlingGame(object):
    def __init__(self):
        self._rolls = []
        self._frame = 0
        self._first = True
        self._extra_rolls = None

    def roll(self, pins):
        if self._game_over():
            raise IndexError("Game Over")
        if not (0 <= pins <= 10):
            raise ValueError("Cannot score {}".format(pins))
        if not self._first and pins + self._rolls[-1] > 10:
            raise ValueError("Cannot score {} and {}".format(self._rolls[-1], pins))

        self._rolls.append(pins)
        self._update_state()

    def score(self):
        if not self._game_over():
            raise IndexError("Game in progress")

        index = frame = total = 0
        rolls = self._rolls
        rolls_len = len(rolls)
        
        while index < rolls_len and frame < 10:
            first = rolls[index]
            second = rolls[index + 1] if index + 1 < rolls_len else 0
            
            if first == 10:  # strike
                third = rolls[index + 2] if index + 2 < rolls_len else 0
                total += 10 + second + third
                index += 1
            elif first + second == 10:  # spare
                third = rolls[index + 2] if index + 2 < rolls_len else 0
                total += 10 + third
                index += 2
            else:  # regular frame
                total += first + second
                index += 2
            
            frame += 1
        
        return total

    def _update_state(self):
        strike = self._rolls[-1] == 10

        self._frame = min(20, self._frame + (2 if strike else 1))
        self._first = True if strike else not self._first

        if self._frame == 20 and self._extra_rolls is None: 
            if strike:
                self._extra_rolls = 2
            elif len(self._rolls) >= 2 and self._rolls[-2] + self._rolls[-1] == 10:
                self._extra_rolls = 1
            else:
                self._extra_rolls = 0
        elif self._extra_rolls is not None:
            self._extra_rolls -= 1

    def _game_over(self):
        return self._frame >= 20 and (self._extra_rolls is None or self._extra_rolls <= 0)

File: src/dataset/javascript/bowling_exercism-ThomasZumsteg/claude/ClaudeSonnet4_bowling_v2.js
class Bowling {
    constructor() {
        this.rolls = [];
        this.frame = 0;
        this.extra = 0;
        this.second = false;
    }

    gameOver() { return this.frame >= 10 && this.extra <= 0; }

    roll(pins) {
        if(this.gameOver()) {
            throw new Error('Cannot roll after game is over');
        } else if(pins < 0) {
            throw new Error('Negative roll is invalid');
        } else if(pins > 10 || (this.second && this.rolls[this.rolls.length - 1] + pins > 10)) {
            throw new Error('Pin count exceeds pins on the lane');
        }

        if(this.frame === 9) {
            if(!this.second && pins === 10) {
                this.extra = 2;
            } else if(this.second && pins + this.rolls[this.rolls.length - 1] === 10) {
                this.extra = 1;
            }
        } else if(this.frame > 9) {
            this.extra--; 
        }

        if(pins === 10) {
            this.frame++;
            this.second = false;
        } else {
            if(this.second) this.frame++;
            this.second = !this.second;
        }
        this.rolls.push(pins);
    }

    score() {
        if(!this.gameOver()) {
            throw new Error('Score cannot be taken until the end of the game');
        }
        let result = 0;
        let i = 0;
        for(let frame = 0; frame < 10; frame++) {
            const roll1 = this.rolls[i];
            const roll2 = this.rolls[i + 1];
            result += roll1 + roll2;
            if(roll1 === 10 || roll1 + roll2 === 10) {
                result += this.rolls[i + 2];
            }
            i += (roll1 === 10 ? 1 : 2);
        }
        return result;
    }
}

module.exports = Bowling;